src/index.ts

```diff
@@ -0,0 +1,414 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) 2014-2015, PhosphorJS Contributors
+|
+| Distributed under the terms of the BSD 3-Clause License.
+|
+| The full license is in the file LICENSE, distributed with this software.
+|----------------------------------------------------------------------------*/
+'use strict';
+
+import {
+  IDisposable
+} from 'phosphor-disposable';
+
+import {
+  Message
+} from 'phosphor-messaging';
+
+import {
+  IChangedArgs, Property
+} from 'phosphor-properties';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  Widget, ResizeMessage
+} from 'phosphor-widget';
+
+
+/**
+ * An interface required for implementing the editor model
+ */
+export
+interface IEditorModel {
+  /**
+   * Updates the buffer stored in the model.
+   *
+   * @param value - A string containing the complete data in the view.
+   */
+  updateBuffer(value: string): void;
+  /**
+   * Saves the current data in the existing file.
+   *
+   * @param data - A string containing the complete data to be saved.
+   */
+  save(data: string): void;
+  /**
+   * Renames the current file.
+   *
+   * @param name - A string denoting the new filename.
+   *
+   * #### Notes
+   * This does *not* remove the existing file, it simply saves a
+   * new file with the same data.
+   */
+  rename(name: string): void;
+  /**
+   * A signal emitted when the editor model state changes.
+   */
+  changed: ISignal<EditorModel, IChangedArgs<any>>
+  /**
+   * The mode of the editor, eg. for code this would be
+   * the language, like 'python' or 'javascript'.
+   */
+  mode: string;
```

Comment:
I chose not to, as codemirror docs here: https://codemirror.net/doc/manual.html stated that you can just use the mode type, instead of the mime type.